File: lib/scripts/migrate_hisn_translations.py
import json
import os

def normalize(text):
    return text.strip()

def main():
    supported_languages = ['bn', 'de', 'es', 'fa', 'fr', 'id', 'ms', 'tr', 'ur', 'zh']
    current_dir = os.getcwd()
    json_dir = os.path.join(current_dir, 'assets', 'json')
    
    print(f"Json Directory: {json_dir}")
    
    husn_en_path = os.path.join(json_dir, 'husn_en.json')
    if not os.path.exists(husn_en_path):
        print(f"Error: husn_en.json not found at {husn_en_path}")
        return

    # Read Template
    with open(husn_en_path, 'r', encoding='utf-8') as f:
        husn_en_json = json.load(f)
    
    # Get first key (usually "English")
    template_key = list(husn_en_json.keys())[0]
    template_list = husn_en_json[template_key]
    
    print(f"Loaded Template with {len(template_list)} chapters.")
    
    # 1. Process Supported Languages
    for lang in supported_languages:
        print(f"\nProcessing Language: {lang}")
        azkar_path = os.path.join(json_dir, f"azkar_{lang}.json")
        out_path = os.path.join(json_dir, f"husn_{lang}.json")
        
        if not os.path.exists(azkar_path):
            print(f"Warning: azkar_{lang}.json not found. Skipping.")
            continue
            
        # Build Map from Azkar
        with open(azkar_path, 'r', encoding='utf-8') as f:
            azkar_list = json.load(f)
            
        translation_map = {}
        category_map = {}
        
        azkar_items = 0
        for category in azkar_list:
            cat_name = category.get('category', '')
            array = category.get('array', [])
            for item in array:
                text = item.get('text', '')
                trans = item.get('translation', '')
                if text:
                    key = normalize(text)
                    if trans:
                        translation_map[key] = trans
                        if cat_name:
                            category_map[key] = cat_name
                azkar_items += 1
        
        print(f"  Loaded {len(translation_map)} translations from {azkar_items} items.")
        
        # Migrate
        new_husn_list = []
        matches = 0
        total_dhikrs = 0
        
        for chapter in template_list:
            new_chapter = chapter.copy()
            text_list = new_chapter.get('TEXT', [])
            new_text_list = []
            
            category_votes = {}
            
            for item in text_list:
                total_dhikrs += 1
                dhikr = item.copy()
                arabic = dhikr.get('ARABIC_TEXT', '')
                key = normalize(arabic)
                
                if key in translation_map:
                    dhikr['TRANSLATED_TEXT'] = translation_map[key]
                    matches += 1
                    
                    if key in category_map:
                        cat = category_map[key]
                        category_votes[cat] = category_votes.get(cat, 0) + 1
                
                new_text_list.append(dhikr)
            
            new_chapter['TEXT'] = new_text_list
            
            # Update Title
            if category_votes:
                best_cat = max(category_votes, key=category_votes.get)
                if best_cat:
                    new_chapter['TITLE'] = best_cat
            
            new_husn_list.append(new_chapter)
            
        # Save
        out_json = {lang: new_husn_list}
        with open(out_path, 'w', encoding='utf-8') as f:
            json.dump(out_json, f, ensure_ascii=False, indent=4)
            
        print(f"  Saved husn_{lang}.json. Match Rate: {matches}/{total_dhikrs}")

    # 2. Process Arabic
    print("\nProcessing Language: ar")
    azkar_ar_path = os.path.join(json_dir, 'azkar.json')
    out_ar_path = os.path.join(json_dir, 'husn_ar.json')
    
    category_map_ar = {}
    if os.path.exists(azkar_ar_path):
        with open(azkar_ar_path, 'r', encoding='utf-8') as f:
            azkar_ar_list = json.load(f)
        for category in azkar_ar_list:
            cat_name = category.get('category', '')
            array = category.get('array', [])
            for item in array:
                text = item.get('text', '')
                if text and cat_name:
                    category_map_ar[normalize(text)] = cat_name
    
    new_husn_list_ar = []
    
    for chapter in template_list:
        new_chapter = chapter.copy()
        text_list = new_chapter.get('TEXT', [])
        new_text_list = []
        
        category_votes = {}
        
        for item in text_list:
            dhikr = item.copy()
            arabic = dhikr.get('ARABIC_TEXT', '')
            key = normalize(arabic)
            
            # Clear translated text for Arabic
            dhikr['TRANSLATED_TEXT'] = ""
            
            if key in category_map_ar:
                cat = category_map_ar[key]
                category_votes[cat] = category_votes.get(cat, 0) + 1
            
            new_text_list.append(dhikr)
        
        new_chapter['TEXT'] = new_text_list
        
        if category_votes:
            best_cat = max(category_votes, key=category_votes.get)
            if best_cat:
                new_chapter['TITLE'] = best_cat
        
        new_husn_list_ar.append(new_chapter)
        
    out_json_ar = {'ar': new_husn_list_ar}
    with open(out_ar_path, 'w', encoding='utf-8') as f:
        json.dump(out_json_ar, f, ensure_ascii=False, indent=4)
        
    print("  Saved husn_ar.json")
    print("\nMigration Complete.")

if __name__ == "__main__":
    main()

File: lib/scripts/migrate_hisn_translations.dart
import 'dart:convert';
import 'dart:io';

Future<void> main() async {
  final supportedLanguages = ['bn', 'de', 'es', 'fa', 'fr', 'id', 'ms', 'tr', 'ur', 'zh'];
  
  // Get absolute path to assets/json
  // Assuming script is run from project root
  final currentDir = Directory.current.path;
  final jsonDir = '$currentDir/assets/json';
  
  print('Json Directory: $jsonDir');

  final husnEnFile = File('$jsonDir/husn_en.json');
  if (!husnEnFile.existsSync()) {
    print('Error: husn_en.json not found at ${husnEnFile.path}');
    return;
  }

  // Read Template (English)
  final husnEnContent = await husnEnFile.readAsString();
  final husnEnJson = jsonDecode(husnEnContent) as Map<String, dynamic>;
  final templateKey = husnEnJson.keys.first; 
  final templateList = husnEnJson[templateKey] as List;

  print('Loaded Template with ${templateList.length} chapters.');

  // Helper to normalize Arabic text
  String normalize(String text) {
    return text.trim();
  }

  // 1. Process Supported Languages
  for (final lang in supportedLanguages) {
    print('\nProcessing Language: $lang');
    final azkarFile = File('$jsonDir/azkar_$lang.json');
    final outFile = File('$jsonDir/husn_$lang.json');

    if (!azkarFile.existsSync()) {
      print('Warning: azkar_$lang.json not found. Skipping.');
      continue;
    }

    // Build Map from Azkar
    final azkarContent = await azkarFile.readAsString();
    final azkarList = jsonDecode(azkarContent) as List;
    
    final Map<String, String> translationMap = {};
    final Map<String, String> categoryMap = {}; // Arabic Text -> Category Name

    int azkarItems = 0;
    for (final category in azkarList) {
      final catName = category['category'] as String? ?? '';
      final array = category['array'] as List? ?? [];
      for (final item in array) {
        final text = item['text'] as String? ?? '';
        final trans = item['translation'] as String? ?? '';
        if (text.isNotEmpty) {
          final key = normalize(text);
          if (trans.isNotEmpty) {
            translationMap[key] = trans;
            if (catName.isNotEmpty) {
              categoryMap[key] = catName;
            }
          }
        }
        azkarItems++;
      }
    }
    print('  Loaded ${translationMap.length} translations from $azkarItems items in azkar_$lang.json');

    // Migrate
    final newHusnList = [];
    int matches = 0;
    int totalDhikrs = 0;

    for (final chapter in templateList) {
      final newChapter = Map<String, dynamic>.from(chapter);
      final textList = (newChapter['TEXT'] as List);
      final newTextList = <Map<String, dynamic>>[];
      
      final categoryVotes = <String, int>{};

      for (final item in textList) {
        final dhikr = item as Map<String, dynamic>;
        totalDhikrs++;
        final arabic = dhikr['ARABIC_TEXT'] as String? ?? '';
        final key = normalize(arabic);

        final newDhikr = Map<String, dynamic>.from(dhikr);
        
        if (translationMap.containsKey(key)) {
          newDhikr['TRANSLATED_TEXT'] = translationMap[key];
          matches++;
          
          if (categoryMap.containsKey(key)) {
            final cat = categoryMap[key]!;
            categoryVotes[cat] = (categoryVotes[cat] ?? 0) + 1;
          }
        }
        
        newTextList.add(newDhikr);
      }
      
      newChapter['TEXT'] = newTextList;

      if (categoryVotes.isNotEmpty) {
        var bestCat = '';
        var maxVotes = 0;
        categoryVotes.forEach((k, v) {
          if (v > maxVotes) {
            maxVotes = v;
            bestCat = k;
          }
        });
        if (bestCat.isNotEmpty) {
           newChapter['TITLE'] = bestCat;
        }
      }

      newHusnList.add(newChapter);
    }

    final outJson = {lang: newHusnList};
    await outFile.writeAsString(jsonEncode(outJson));
    print('  Saved husn_$lang.json. Match Rate: $matches / $totalDhikrs (${(matches/totalDhikrs*100).toStringAsFixed(1)}%)');
  }

  // 2. Process Arabic (husn_ar.json)
  print('\nProcessing Language: ar');
  final outFile = File('$jsonDir/husn_ar.json');
  final azkarFile = File('$jsonDir/azkar.json');
  
  Map<String, String> categoryMapAr = {};

  if (azkarFile.existsSync()) {
      final azkarContent = await azkarFile.readAsString();
      final azkarList = jsonDecode(azkarContent) as List;
      for (final category in azkarList) {
          final catName = category['category'] as String? ?? '';
          final array = category['array'] as List? ?? [];
          for (final item in array) {
             final text = item['text'] as String? ?? '';
             // Use text as key to identify category
             if (text.isNotEmpty && catName.isNotEmpty) {
               categoryMapAr[normalize(text)] = catName;
             }
          }
      }
      print('  Loaded Arabic categories from azkar.json');
  } else {
      print('  Warning: azkar.json not found. Retaining English titles.');
  }

  final newHusnListAr = [];
  int totalDhikrsAr = 0;
  
  for (final chapter in templateList) {
    final newChapter = Map<String, dynamic>.from(chapter);
    final textList = (newChapter['TEXT'] as List);
    final newTextList = <Map<String, dynamic>>[];
    
    final categoryVotes = <String, int>{};

    for (final item in textList) {
      final dhikr = item as Map<String, dynamic>;
      totalDhikrsAr++;
      final arabic = dhikr['ARABIC_TEXT'] as String? ?? '';
      final key = normalize(arabic);

      final newDhikr = Map<String, dynamic>.from(dhikr);
      
      // Clear Translated Text for Arabic mode
      newDhikr['TRANSLATED_TEXT'] = ""; 
      
      if (categoryMapAr.containsKey(key)) {
            final cat = categoryMapAr[key]!;
            categoryVotes[cat] = (categoryVotes[cat] ?? 0) + 1;
      }
      
      newTextList.add(newDhikr);
    }
    
    newChapter['TEXT'] = newTextList;
    
    // Update Title to Arabic
    if (categoryVotes.isNotEmpty) {
        var bestCat = '';
        var maxVotes = 0;
        categoryVotes.forEach((k, v) {
            if (v > maxVotes) {
            maxVotes = v;
            bestCat = k;
            }
        });
        if (bestCat.isNotEmpty) {
            newChapter['TITLE'] = bestCat;
        }
    }

    newHusnListAr.add(newChapter);
  }

  final outJsonAr = {'ar': newHusnListAr};
  await outFile.writeAsString(jsonEncode(outJsonAr));
  print('  Saved husn_ar.json with cleared translations and mapped titles.');
  print('\nMigration Complete.');
}
